src/hermit.rs

```diff
@@ -0,0 +1,21 @@
+use crate::Error;
+use core::{cmp::min, mem::MaybeUninit, num::NonZeroU32};
+
+extern "C" {
+    fn sys_read_entropy(buffer: *mut u8, length: usize, flags: u32) -> isize;
+}
+
+pub fn getrandom_inner(mut dest: &mut [MaybeUninit<u8>]) -> Result<(), Error> {
+    while !dest.is_empty() {
+        let res = unsafe { sys_read_entropy(dest.as_mut_ptr() as *mut u8, dest.len(), 0) };
+        if res < 0 {
+            // SAFETY: all Hermit error codes use i32 under the hood:
+            // https://github.com/hermitcore/libhermit-rs/blob/master/src/errno.rs
+            let code = unsafe { NonZeroU32::new_unchecked((-res) as u32) };
```

Comment:
For this, could we avoid unsafe code by doing:
```rust
if res < 0 {
    return Err(match NonZeroU32::new((-res) as u32) {
        Some(code) => code.into(),
        None => Error::ERRNO_NOT_POSITIVE,
    });
}
```
which is sort of similar to code we use elsewhere, and avoids and unsafe or panicking code.

Reply:
I think `res < 0` followed by `ERRNO_NOT_POSITIVE` would be... a bit confusing. The reason why we can't simply `.unwrap()` here is because the syscall returns `isize`, thus on 64-bit platforms it's theoretically  possible to get a negative number with lower 32 bits equal to zero.

Reply:
That's very fair. Maybe we can just have an unwrap then instead of an unsafe block. The panic branch would only be in the error handling, so it doesn't seem like an issue.

Reply:
I think that using `ERRNO_NOT_POSITIVE` as an unreachable fallback constant is better than introduction of an unreachable panic branch. Maybe we should introduce `UNSPECIFIED_ERROR` as `internal_error(2)`?

Reply:
What if we just added the following method to `Error`:
```rust
pub fn from_raw_os_error(errno: i32) -> Self {
    if errno > 0 {
        Error::from(NonZeroU32::new(errno as u32).unwrap())
    } else {
        Error::ERRNO_NOT_POSITIVE
    }
}
```

And in `hermit.rs` we just had:
```
if res < 0 {
    return Err(Error::from_raw_os_error(-res as i32));
}
```

Reply:
Actually, I'm fine leaving this as-is for now. I've got a followup PR that might make this sort of thing easier going forward.